locale/src/locale.cc

```diff
@@ -0,0 +1,72 @@
+#include <leatherman/locale/locale.hpp>
+#include <map>
+
+// boost includes are not always warning-clean. Disable warnings that
+// cause problems before including the headers, then re-enable the warnings.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstrict-aliasing"
+#include <boost/locale.hpp>
+#include <boost/thread/shared_mutex.hpp>
+#include <boost/thread/lock_guard.hpp>
+#include <boost/thread/shared_lock_guard.hpp>
+#pragma GCC diagnostic pop
+
+
+namespace leatherman { namespace locale {
+
+    static boost::shared_mutex g_locales_mutex;
+    static std::map<std::string, std::locale> g_locales;
+
+    const std::locale& get_locale(std::string const& id, std::string const& domain, std::vector<std::string> const& paths)
+    {
+        {
+            // Allow multiple readers of the map, but only when there are no writers.
+            boost::shared_lock_guard<boost::shared_mutex> guard(g_locales_mutex);
+            auto it = g_locales.find(domain);
+            if (it != g_locales.end()) {
+                return it->second;
+            }
+        }
+
+        // The system default locale is set with id == "", except on Windows boost::locale's
+        // generator uses a compatible UTF-8 equivalent. Using boost results in UTF-8 being
+        // the default on all platforms.
+        boost::locale::generator gen;
+#ifdef LEATHERMAN_LOCALE_INSTALL
+        // Setup so we can find installed locales.
+        gen.add_messages_path(LEATHERMAN_LOCALE_INSTALL);
+#endif
+        for (auto&& path : paths) {
+            gen.add_messages_path(path);
```

Comment:
Since `paths` is obviously const and the intent wasn't to move the vector's elements, `auto&` is fine for constant references: the type deduced is `T const`, making it `T const&`, as intended.